tutorials/preprocessing/plot_55_setting_eeg_reference.py

```diff
@@ -0,0 +1,122 @@
+# :meth:`~mne.io.Raw.set_eeg_reference` method. Of course, you can only set
+# the reference to a specific (set of) electrode(s) if those electrodes are
+# present in the :class:`~mne.io.Raw` file; the :ref:`example data
+# <sample-dataset>` has numeric EEG channel names (``EEG 001``, etc) so the
+# double mastoid and earlobe examples are commented out here:
+
+# use average of mastoid channels as reference
+# raw.set_eeg_reference(ref_channels=['M1', 'M2'])
+
+# use a single channel reference (left earlobe)
+# raw.set_eeg_reference(ref_channels=['A1'])
+
+# use the average of all channels as reference
+raw_avg_ref = raw.copy().set_eeg_reference(ref_channels='average')
+
+###############################################################################
+# When setting ``ref_channels='average'``, MNE-Python will automatically
+# exclude any channels listed in ``raw.info['bads']`` from contributing to the
+# average reference.
+#
+# Creating the average reference as a projector
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# If using an average reference, it is possible to create the reference as a
+# :term:`projector` rather than subtracting the reference from the data
+# immediately, by specifying ``projection=True``:
+
+raw.set_eeg_reference('average', projection=True)
+print(raw.info['projs'])
+
+###############################################################################
+# Creating the average reference as a projector has a couple of advantages:
+#
+# 1. it is possible to turn projectors on or off when plotting, so it is easy
+#    to visualize the effect that the reference has on the data
+#
+# 2. if there are other unapplied projectors affecting the EEG channels (such
+#    as SSP projectors for removing heartbeat or blink artifacts), EEG
+#    referencing cannot be performed until those projectors are either applied
+#    or removed; adding the EEG reference as a projector is not subject to that
+#    constraint. (The reason this wasn't a problem when we created
+#    ``raw_avg_ref`` above is that the projectors )
+
+eeg_picks = mne.pick_types(raw.info, meg=False, eeg=True)
+for title, proj in zip(['No reference', 'Average reference'], [False, True]):
+    fig = raw.plot(proj=proj, order=eeg_picks, n_channels=len(eeg_picks))
+    # make room for title
+    fig.subplots_adjust(top=0.94)
+    fig.suptitle(title, size='xx-large', weight='bold')
+
+###############################################################################
+# .. warning::
+#
+#     MNE-Python will automatically apply an average reference if EEG channels
+#     are present and no reference strategy is specified. Thus if you are
+#     loading partially-preprocessed data that has already had a reference
+#     applied, you should set the reference to an empty list
+#     (``raw.set_eeg_reference(ref_channels=[])``) to prevent MNE-Python from
+#     subtracting a second average reference signal from your data.
```

Comment:
I this really true? I've never seen this in practice, after loading MNE didn't change the reference until I explicitly called `raw.set_eeg_reference`.

Reply:
good question. The docstring says so:

> By default, MNE-Python will automatically re-reference the EEG signal
to use an average reference...

but digging into the code a little it seems like it only ever happens during inverse imaging / covariance computation?  And at that point it appears to be *mandatory*, so the implication that it's OK to set `ref_channels=[]` and still do inverse imaging is maybe not correct...  If anyone knows the answer off the top of their head it would save me some time digging deeper through all the code paths...

Reply:
This was exactly my understanding, but I haven't checked recently.

Reply:
Yeah it should really be required by our codebase for source localization. The docstring should be updated

Reply:
The more I dig into this the more confused I am.  Here are, I think, the possibilities:

1. user never specifies EEG reference → inverse imaging → average ref happens automatically
2. user calls `set_eeg_reference([])` → inverse imaging → works
3. user calls `set_eeg_reference('average', projection=True)` → inverse imaging → works
4. user calls `set_eeg_reference('average')` → inverse imaging → ValueError
5. user calls `set_eeg_reference(['Cz'])` → inverse imaging → ValueError
6. user calls `add_reference_channels('custom ref channel name(s)')` → inverse imaging → ValueError

I would have thought #4 should work, and I don't understand the difference (in channel data values) between #2 and #6.

Reply:
#2 means: the EEG data already has an appropriate reference, don't re-reference the data, I know what I'm doing. MNE-Python doesn't know what the reference is, but will obey the user in this case.

#4 should probably be made to work,

Reply:
No wait, #4 cannot work, because inverse modeling requires that the EEG signal has zero mean. The reference needs to be done with a projector, so the zero mean requirement can be guaranteed, even after picking channels.